### DISCUSSION TITLE
How to use Groq with Browser-use

### DISCUSSION THREAD

**Discussion by Sri-Krishna-V** [browser-use/browser-use#113] [Q&A]
How should I use Browser-use with Groq? 
Loaded my groq api key using .env

Here's my code:
```python
from langchain_groq import ChatGroq
from browser_use import Agent
import asyncio
from dotenv import load_dotenv
import os


load_dotenv()

llm = ChatGroq(
    model_name="llama-3.3-70b-versatile",
)

async def main():
    agent = Agent(
        task="Find a one-way flight from Coimbatore to Kolkata on 12 January 2025 on Google Flights. Return me the cheapest option.",
        llm = llm
    )
    result = await agent.run()
    print(result)

asyncio.run(main())
``` 

Error:

```powershell
INFO     [browser_use] BrowserUse logging setup complete with level info
INFO     [root] Anonymized telemetry enabled. See https://github.com/gregpr07/browser-use for more information.
INFO     [agent] 🚀 Starting task: Find a one-way flight from Coimbatore to Kolkata on 12 January 2025 on Google Flights. Return me the cheapest option.
INFO     [agent] 
📍 Step 1
ERROR    [agent] ❌ Result failed 1/5 times:
 Error code: 400 - {'error': {'message': 'message[2].content must be a string', 'type': 'invalid_request_error'}}
INFO     [agent] 
📍 Step 1
ERROR    [agent] ❌ Result failed 2/5 times:
 Error code: 400 - {'error': {'message': 'message[2].content must be a string', 'type': 'invalid_request_error'}}
INFO     [agent] 
📍 Step 1
ERROR    [agent] ❌ Result failed 3/5 times:
 Error code: 400 - {'error': {'message': 'message[2].content must be a string', 'type': 'invalid_request_error'}}
INFO     [agent] 
📍 Step 1
ERROR    [agent] ❌ Result failed 4/5 times:
 Error code: 400 - {'error': {'message': 'message[2].content must be a string', 'type': 'invalid_request_error'}}
INFO     [agent] 
📍 Step 1
ERROR    [agent] ❌ Result failed 5/5 times:
 Error code: 400 - {'error': {'message': 'message[2].content must be a string', 'type': 'invalid_request_error'}}
ERROR    [agent] ❌ Stopping due to 5 consecutive failures
AgentHistoryList(all_results=[ActionResult(is_done=False, extracted_content=None, error="Error code: 400 - {'error': {'message': 'message[2].content must be a string', 'type': 'invalid_request_error'}}", include_in_memory=True), ActionResult(is_done=False, extracted_content=None, error="Error code: 400 - {'error': {'message': 'message[2].content must be a string', 'type': 'invalid_request_error'}}", include_in_memory=True), ActionResult(is_done=False, extracted_content=None, error="Error code: 400 - {'error': {'message': 'message[2].content must be a string', 'type': 'invalid_request_error'}}", include_in_memory=True), ActionResult(is_done=False, extracted_content=None, error="Error code: 400 - {'error': {'message': 'message[2].content must be a string', 'type': 'invalid_request_error'}}", include_in_memory=True), ActionResult(is_done=False, extracted_content=None, error="Error code: 400 - {'error': {'message': 'message[2].content must be a string', 'type': 'invalid_request_error'}}", include_in_memory=True)], all_model_outputs=[])
```

**Comment by rhyliieee** [comment]
Following, since I face the same problem.

**Reply by rhyliieee** [reply]
Resolved mine, you have to change to a multi-modal model for it to work. But after changing mine to 3.2 90B vision, It shows a message saying `{'error': {'message': 'Too many images provided.  This model supports up to 1 images', 'type': 'invalid_request_error'}}`

**Reply by Sri-Krishna-V** [reply]
This is the problem i guess


Llama 3.2 90B Vision (Preview)
Model ID
`llama-3.2-90b-vision-preview`
Description
A powerful multimodal model capable of processing both text and image inputs that supports multilingual, multi-turn conversations, tool use, and JSON mode.
Context Window
8,192 tokens
Preview Model
Currently in preview and should be used for experimentation.
Image Size Limit
Maximum allowed size for a request containing an image URL as input is 20MB. Requests larger than this limit will return a 400 error.
Request Size Limit (Base64 Encoded Images)
Maximum allowed size for a request containing a base64 encoded image is 4MB. Requests larger than this limit will return a 413 error.
```python
Single Image per Request
Only one image can be processed per request in the preview release. Requests with multiple images will return a 400 error.
``` 
System Prompt
Does not support system prompts and images in the same request.

Maybe llama3-groq-70b-8192-tool-use-preview this could help?

**Reply by moodpanda** [reply]
> Following, since I face the same problem.

I tried to change the system prompt to user prompt it work but my problem now is the token size limit it timeout me after some steps
![image](https://github.com/user-attachments/assets/bdb5a87a-57ed-433d-bbad-7eded9263552)

![image](https://github.com/user-attachments/assets/5b43756d-5399-47b3-84c5-3a27b267ada9)

error:
ERROR    [agent] ❌ Result failed 1/3 times:
 Error code: 400 - {'error': {'message': 'Please reduce the length of the messages or completion.', 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}

**Comment by ameyasutar1** [comment]
Have you tried Using gemini 2.0 flash Model ?

**Comment by jingchang0623-crypto** [comment]
Quick tip: if you want to use Groq (or any other LLM provider) with browser-use, the cleanest approach is through **LiteLLM** as a proxy. It gives you a unified OpenAI-compatible interface for 100+ providers.

## Setup
```bash
pip install litellm
```

```python
from browser_use import Agent
from langchain_openai import ChatOpenAI
import asyncio

# LiteLLM proxy setup
llm = ChatOpenAI(
    model="groq/llama-3.3-70b-versatile",
    api_key="your-groq-key",
    base_url="https://api.litellm.ai/v1"
)

async def main():
    agent = Agent(
        task="Find cheapest flight from Coimbatore to Kolkata on Jan 12",
        llm=llm,
        max_steps=20,
    )
    result = await agent.run()
    print(result)

asyncio.run(main())
```

## Why LiteLLM?
- Drop-in replacement, no code changes needed for other providers
- Automatic retry and fallback logic
- Cost tracking across providers
- Works with Groq, Together, Anthropic, Gemini, etc.

## Speed Note ⚡
Groq is insanely fast for inference, but browser-use does a lot of waiting for page loads anyway. If speed is your goal, consider running browser-use with multiple parallel agents instead of optimizing for the fastest LLM.

Also saw discussion #31 about LiteLLM testing — worth checking that thread too for more details!

**Comment by Ghraven** [comment]
For current browser-use versions, I would avoid passing `langchain_groq.ChatGroq` directly here and use browser-use's native Groq wrapper instead:

```python
import asyncio
from browser_use import Agent
from browser_use.llm import ChatGroq

llm = ChatGroq(
    model="meta-llama/llama-4-maverick-17b-128e-instruct",
    temperature=0.0,
    # api_key="...",  # optional if GROQ_API_KEY is already in your env
)

async def main():
    agent = Agent(
        task="Find a one-way flight from Coimbatore to Kolkata on Google Flights and return the cheapest option.",
        llm=llm,
    )
    await agent.run()

asyncio.run(main())
```

There is also an example in the repo under `examples/models/llama4-groq.py` / `examples/getting_started/05_fast_agent.py`.

One practical caveat: not every Groq-hosted model is equally good at browser-use's structured/tool-calling loop. If you still see schema/content errors, try one of the models browser-use marks as verified for Groq, such as `meta-llama/llama-4-maverick-17b-128e-instruct`, and keep temperature low.

**Comment by Ghraven** [comment]
I am not a maintainer, but I think the issue is that this example is using `langchain_groq.ChatGroq` directly. Browser-use can pass multimodal/content-part messages during a browser step, and Groq's LangChain wrapper may reject that with `message[2].content must be a string`.

For current browser-use, I would try the built-in Groq wrapper instead:

```python
import asyncio
from dotenv import load_dotenv
from browser_use import Agent, ChatGroq

load_dotenv()

llm = ChatGroq(
    model="meta-llama/llama-4-maverick-17b-128e-instruct",
    temperature=0,
)

async def main():
    agent = Agent(
        task="Find a one-way flight from Coimbatore to Kolkata on a future date on Google Flights. Return the cheapest option.",
        llm=llm,
    )
    result = await agent.run()
    print(result)

asyncio.run(main())
```

If you want to keep using `langchain_groq`, another thing to test is `Agent(..., use_vision=False)`, because that keeps the model input text-only. The tradeoff is that browser navigation may be less capable without screenshots.

One small note: the date in the original task, 12 January 2025, is already in the past, so Google Flights may also behave oddly even after the model wrapper issue is fixed.
